Cython/Includes/cpython/complex.pxd

```diff
@@ -12,14 +13,20 @@ cdef extern from "Python.h":
     # PyComplexObject
     # This subtype of PyObject represents a Python complex number object.
 
-    ctypedef class __builtin__.complex [object PyComplexObject]:
+    # It actually isn't even opaque in the limited API - it just isn't present. However, "opaque"
+    # removes enough code that it isn't a compile error
+    ctypedef class __builtin__.complex [object PyComplexObject, check_size opaque_in_limited_api]:
         cdef Py_complex cval
 
+        # unavailable in limited API
         @property
+        @_cython.c_compile_guard("!CYTHON_COMPILING_IN_LIMITED_API")
```

Comment:
Maybe(!), we should ignore `inline` methods/properties defined on builtin types in the Limited API? I'm not sure, and it's somewhat of a stretch in motivating it, but they seem a bit more likely to get in the way on a cross-version Stable ABI build (although that's not entirely the only reason for the Limited API), and also more likely to not work at all, as in this case. The way I read them, they are really a performance feature. If they only override existing Python API (and we could even check that and restrict their removal to this case), then there's not much of a reason for them in the Limited API.

So, check if they override existing attributes/methods and then discard them in the Limited API?

Reply:
I'm not quite sure about this.

> The way I read them, they [cdef inline properties] are really a performance feature. If they only override existing Python API (and we could even check that and restrict their removal to this case), then there's not much of a reason for them in the Limited API.

Yes. The difficulty is disabling them in the Limited API without blocking access to the Python API. I can't see a good way of doing that.

------------------------------------

I've removed the changes in datetime.pxd since almost the whole thing was disabled in the Limited API so I think it's best to have a warning and give up.

That leaves:

* `real`/`imag` in `complex.pxd`
* A couple of inline module-scoped functions in `contextvar.pxd`.

I think the module-scoped functions mean "ignore inline methods/properties defined on builtin types" wouldn't really work.

So the options are:
* Do something similar to what I've done.
* Rewrite them to work with the Limited API. For the complex ones, it'd be hard to make it truly `noexcept`. The contextvar ones would just have to use the Python API

I'm not sure which is best.

Reply:
@scoder can you have another look at this?

Reply:
> I've removed the changes in datetime.pxd since almost the whole thing was disabled in the Limited API so I think it's best to have a warning and give up.

A warning seems ok here. It might be annoying for some, but then the best solution is to extract the C-API usages into a separate module that gets optionally built and imported when not using the Limited API. So there are still ways to write code for both targets.

> A couple of inline module-scoped functions in `contextvar.pxd`.

There's already a verbatim code section at the top of the file that (stupidly) defines `PyContextVar_Get()` in older Python versions that lack it. We could reuse that for the Limited API to make it compile and add a warning as a comment that the contextvars C-API simply isn't available there. That would at least allow users to write code that uses the C-API if available and fall back to `import contextvars` otherwise.

We could also turn the macro into an inline function and then issue an actual Python warning if it gets called. That would help making users more aware of a problem in their code.

> real/imag in complex.pxd

Since these are Cython inline properties (and not Python properties), we could use a C inline function for the struct access that calls `PyObject_GetAttr()` in the Limited API.

The `noexcept` gets in the way then, though… but since failures should not occur in practice, we could still keep it and live with the "unraisable exception"…